---
title: 'Introduction To R: Part Two - Worksheet'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This worksheet accompanies the learning material for Introduction to R: Part Two. 
Complete the following questions, solutions will be released on Thursday April 30th following the Zoom conference.

## Part One
1. Create a vector of the first 16 prime numbers.

2. Use this vector to create a 4x4 matrix called ```my_mat``` using default parameters.

3. Add the numbers 1-4 as a new row to the matrix.

4. Replace the entry at the position 2,3 with 3.

5. Write a loop that checks if the entries of the fifth row are even. If they are print "You found me", otherwise print "Try again"

6. Remove the third row.

7. Print out the final matrix.

## Part Two
You have been asked to analyse the results of a differential expression analysis using DESEQ2<sup>1</sup>. You have been supplied with ```results.txt```, the output of this analysis. This includes:
- baseMean: The average of the normalised counts per gene across all samples
- log2FoldChange: Gives an idea of change in expression due to a test condition with respect to control
- lfcSE: Standard error of log2 fold change
- pval: P value, the result of a hypothesis test to test whether the change in expression can be attributed to the test condition
- padj: P value adjusted for multiple testing

1. Read in the file and assign the variable name  ```de```.

2. Head the first 10 rows.

3. Tail the last 8 rows of columns 2 to 3.

4. Identify the number of columns and rows.

5. Change the name of the third column to "L2FC".

6. Extract columns 1:3, rows 1:100 and save as ```sliced_df```. Head the result.

7. Are there any duplicates? How many? If their are duplicates, remove them, then set the first column as rownames and remove the first column. (Continue with dataframe from 5. ```de``` NOT ```sliced_df```)

8. Are there any missing values? If so, remove rows containing them.

9. The information regarding the following gene was mistakenly left out of the dataset, correct this mistake.
   - ENSMUSG00000039287 570.805924  -0.4648999  0.09045180   2.751007e-07    6.393878e-04
   
10. The Wald statistic is generated by dividing log2 fold change by lfcSE and is used to generate the p value. It is missing from this dataset. Add a column that contains a Wald Statistic for each gene and call it ```stat```.

11. Using both (a) summary and (b) a for loop, find the mean of each column, storing the loop output in a vector.

12. Print the results of 10.

13. How many genes are less than 0.05 in both the pval and padj columns?

14. Use a for loop and an if else statement to fill a new column called ```Significance```. If the padj column is less than 0.05 the Significance value should be "Sig" otherwise it should be "Not Sig".

15. There is a cleaner and easier way to do this. Repeat the exercise in 13 without loops or ifelse, except this time add values to the new column ```Expression```. If the value in the ```L2FC``` column is > 1 or less than -1, the corresponding value in ```Expression``` should be "big_change", otherwise it should be "little_change". Hint: Read through section 4.1 - 4.5 of the tutorial for inspiration!

16. Find the dimensions of the dataframe that satisfies the following conditions:
    pvalue < 0.05 
    OR
    L2FC > 1
    
17. Replace ENSMUSG in the rownames with MOUSE.

18. Find the rownumbers of all rows whose rownames have the following pattern: "126".

19. Remove all genes whose expression values < -1 or > 1 and whose adjusted p value is less than 0.05. Call the new dataframe ```interesting_res```.

20. Order by adjusted p value in increasing order of signifcance. Call the new dataframe ```interesting_res_ordered```.

21. Write out the final dataset to a comma separated file called ```DE_RESULTS.csv```. See if you can read it back in without issues.

1. Love, M.I., Huber, W., Anders, S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2 Genome Biology 15(12):550 (2014)
